edrumulus.ino

```diff
@@ -199,6 +199,15 @@ void loop()
           midi_note = edrumulus.get_midi_note_open ( pad_idx );
         }
       }
+      // special case: MPS 750x ride bell heuristic
```

Comment:
I do not want to have pad-specific fixes in the .ino file. One possibility would be to fix this issue with the pad in your xtalk Python script. Would that be an option for you to put the modification of the MIDI note in that script instead of the Edrumulus code?

Reply:
> I do not want to have pad-specific fixes in the .ino file.

Ok, but where should it go then?

> One possibility would be to fix this issue with the pad in your xtalk Python script. Would that be an option for you to put the modification of the MIDI note in that script instead of the Edrumulus code?

No, it doesn't belong to a cross-talk cancellation filter either imo.

Reply:
You play one instrument (bell) but get the sound of another one (rim). In my opinion, this is crosstalk. But if you do not want to correct that in your script, then that's ok with me.

Reply:
Maybe in your ride cymbal there is the same schematics as [ignotus666](https://github.com/ignotus666) describes in [his post](https://github.com/corrados/edrumulus/issues/112#issuecomment-1802744950): "_Piezo/switch/switch hi hat and ride. Same as the crashes + bell switch. Here the edge switch goes through a 10k resistor to distinguish it from the bell switch (both switches go to ring on the jack so all three zones just use a single stereo jack, like Yamaha does with their pads)._"

Reply:
Imo cross-talk are unintended triggers caused by intended triggers. The unintended triggers are supposed to be cancelled. Here however we have an intended trigger (low volume rim shot), which I just change to another intended trigger (bell). In particular I do not want to cancel anything at all.

This is a workaround to make the bell _somewhat_ work. A proper implementation would certainly look different and probably it's the piezo/switch/switch  resistor trick @ignotus666 mentioned (in fact I measured a 5k resistance difference on the two switches).

I think to properly identify which switch is in use, one would have to measure the resistance in place during the switch trigger (if it changes by some kOhm, it's the other trigger), but that would be a much larger implementation project to support these kind of pads in general for edrumulus, which is out of the scope of this PR.

Reply:
> in fact I measured a 5k resistance difference on the two switches

Very interesting. Is it possible by changing the rim shot threshold to distinguish between these two? E.g., change the threshold that only one works (either rim or bell) and then change the threshold so that both can be detected correctly. If that works, it would be possible to introduce a second threshold to distinguish between both.

Reply:
I think you misunderstood my comment. What you are doing is to use the MIDI velocity as the indicator which part of the instrument was played. What I proposed is completely independent of the MIDI velocity. My proposal is all about the detection of a rim shot. Is it detected or not (i.e., true or false).

If you can set the rim shot threshold so that it only triggers (i.e., returns a "is a rim shot") for one of the instrument parts (rim or bell) and if you find another rim shot threshold where both instrument parts are detected correctly, then you can distinguish between them without having to use the MIDI velocity at all.

Reply:
I tested this, but unfortunately couldn't find any such parameter. Whatever I chose, both rim switches went quiet at the same time.

Also, I simulated this in LTSpice: With Rs=10k, Rp=10k I obtain an effective Rs of 41-46k (the input impedance is either 31k or 36k for edge/bell). That results in a voltage difference of 15mV. For Rs=100k, Rp=22k I obtain roughly 4mV difference. Choosing Rs=10k, Rp=43k (which I can't use due to the current that is drawn) would result in a theoretical 40mV difference. In theory all of these can be noticed from the 3,3V/2^10=3mV effective resolution ADC, but practice apparently looks different...

Should I remove the lines above?

Reply:
That's really unfortunate...
You cannot use all bits the ADC supports because of the conversion noise. So, 4 mV is too little difference. One would have to change the front end schematics just for such a pad which is not good.

> Should I remove the lines above?

Yes, that would my plan. We have to get your Pull Request on main so that it is not necessary for you to fix the incompatibilities to main all the time. And at the same time I do not want that code on main.
So, as a temporary workaround I would suggest that you remove these lines and I will merge your code to main. Then you can re-introduce this piece of code on a side branch in your repo clone.

Later on I would really like to have a good solution for such pads. But, unfortunately, right now I am running out of ideas. Let's postpone this a bit...